### ISSUE TITLE
Parser TCK returns different representations for a type when its referenced using type vs using schema

### ISSUE THREAD

**Issue by juancoen** [raml-org/raml-js-parser-2#102]
RAML example

```yaml
#%RAML 1.0
title: API with Types
baseUri: /media

schemas:
  - Player: |
      {
        "type": "object",
        "properties": [
          {
            "name": "string"
          }
        ]
      }
types:
  Person:
    type: object
    properties:
      firstname:
        type: string
        description: test
      lastname: string
      addresses: string[]
      age: number
  User:
    type: Person
    properties:
      firstname:
        type: string
        required: false
      id: number
      department: Department[]
  Developer:
    type: User
    properties:
      github: string
  Department:
    type: object
    properties:
      name: string
    example:
      name: Engineering
  CEO:
    type: User | Developer
  Cleaner: |
    {
      "type": "object",
      "properties": [
        {
          "name": "string"
        }
      ]
    }

/users:
  get:
    body:
      application/json:
        type: object
        properties:
          developer: string
          test: string
          animal: string
          car: string

    responses:
      199:
        body:
          application/json:
            type: object
            properties:
              developer: string
              test: string
              animal: string
              car: string
      200:
        body:
          application/json:
            type: Person
      201:
        body:
          application/json:
            schema: Cleaner
      202:
        body:
          application/json:
            schema: Player
      203:
        body:
          application/json:
            schema: Department
      204:
        body:
          application/json:
            type: |
              {
                "type": "object",
                "properties": [
                  {
                    "name": "string"
                  }
                ]
              }
      205:
        body:
          application/json:
            schema: |
              {
                "type": "object",
                "properties": [
                  {
                    "name": "string"
                  }
                ]
              }
```

in response 200, the property `type` is used to reference the Person type, the parser TCK returns:

```json
{
  "application/json": {
    "displayName": "application/json",
    "name": "application/json",
    "type": [
      "Person"
    ],
    "repeat": false,
    "required": true,
    "__METADATA__": {
      "primitiveValuesMeta": {
        "displayName": {
          "calculated": true
        },
        "repeat": {
          "insertedAsDefault": true
        },
        "required": {
          "insertedAsDefault": true
        }
      }
    }
  },
  "currentType": "application/json"
}
```

in response 201, the property `schema` is used to reference the Cleaner type, the parser TCK returns:

```json
{
  "application/json": {
    "displayName": "application/json",
    "name": "application/json",
    "schema": "Cleaner",
    "type": [
      "object"
    ],
    "repeat": false,
    "required": true,
    "schemaContent": "Cleaner",
    "__METADATA__": {
      "primitiveValuesMeta": {
        "displayName": {
          "calculated": true
        },
        "type": {
          "insertedAsDefault": true
        },
        "repeat": {
          "insertedAsDefault": true
        },
        "required": {
          "insertedAsDefault": true
        }
      }
    }
  },
  "currentType": "application/json"
}
```

Being `type` and `schema` synonymous, there should be no difference in the way they are returned to the parser client.

Thanks!

**Comment by sichvoge** [comment]
Do you mean that in one it basically return an array (in case of `type`) and in the other a simple string (in case of `schema`)?

I'd also add that `schema` and `type` are mutual exclusive according to the spec, so the resulting JSON should not contain both.

**Comment by juancoen** [comment]
What I'd expect for Cleaner is:
```
{
  "application/json": {
    "displayName": "application/json",
    "name": "application/json",
    "type": [
      "Cleaner"
    ],
    "repeat": false,
    "required": true,
    "__METADATA__": {
      "primitiveValuesMeta": {
        "displayName": {
          "calculated": true
        },
        "repeat": {
          "insertedAsDefault": true
        },
        "required": {
          "insertedAsDefault": true
        }
      }
    }
  },
  "currentType": "application/json"
}
```

**Comment by ddenisenko** [comment]
Well, while type and schema are synonyms, they are not the same word. 

1) Users of AST should be able to differentiate what was typed in RAML, so you see "schema" for schema in AST and "type" for type. 

2) Schemas are also types, so you see "type" there too, but in addition you get a special schema-only helper: `schemaContent()` method  as it is supposed that users may probably want to have an easy access to the schema text. The content of "schemaContent" JSON field here should contains the schema text instead of schema name, its a bug.

**Comment by sichvoge** [comment]
@ddenisenko can you create another issue for that?

in terms of having schema and type both in there at the same time; imho - thats a bug too. the spec explicitly says that both are mutual exclusive and if your arguments against Juan's suggestion is the spec, we should take that also as a bug.

**Comment by juancoen** [comment]
So basically, this is pseudo-code for the algorithm needed to handle a body node just for displaying it:
(`display as string` and `display as type` are known functions)

```
if (node has schema property and it is a key in schemas top section) =>
  display the entry in schemas top section as string

else if (node has schema property and it is a key in types top section) =>
   if (entry has "schemaContent" property) =>
      display "schemaContent" property as string
   else =>
      display as type

else if (node has type property and it is a key in schemas top section) =>
  display the entry in schemas top section as string

else if (node has type property and it is a key in types top section) =>
  if (entry has "schemaContent" property) =>
    display "schemaContent" property as string
  else =>
    display as type

else if (node has type property and properties property) =>
  display as type

else =>
  display as string
```

First, will this cover all possible cases?

Second, does anyone else find this *very complex*?

**Comment by juancoen** [comment]
I found another thing that I don't understand. In this case:
```
      203:
        body:
          application/json:
            schema: Department
```

`Department` is defined as `schema` but it's defined as a `type` in the `types` section. The parser returns:

```
"203": {
              "code": "203",
              "body": {
                "application/json": {
                  "displayName": "application/json",
                  "name": "application/json",
                  "schema": "Department",
                  "type": [
                    "object"
                  ],
                  "repeat": false,
                  "required": true,
                  "schemaContent": "Department"
                }
              }
            },
```
Why is the `schemaContent` property there if it's not a `schema`?

**Comment by ddenisenko** [comment]
>Why is the schemaContent property there if it's not a schema?
Because schemaContents appears when you use "schema" keyword.

So, I analyzed the samples again, and went through the algorithm, Juan provided.

It looks like our mixture of schemas and types comes to the point of referencing types in `schema` keyword. I personally dislike this and think we should not be mixing those two to that extent, we've got `schema` keyword as a legacy support, after all. I would left `schema` only being able to reference schemas defined using `schemas` keyword, and only allow defining actual schemas under `schemas` keyword.

But anyway, if we are forced to live with this mixture, then the following should be done:

**Changes:**

1) `schema()` method of the parser (and consequently) the `schema` property in JSON should be modified to return an array of strings to be equal of what `type()` returns. Just because someone could be putting an array of types there.
If we go further along this way, I could be imagining someone writing the following code:
```RAML
...
  body:
    application/json:
      schema: [SomeObjectTypeDefinedUsingTypesProperty, AnotherType]
      properties:
         name: string
         ...
```
But this is a topic for another discussion.

2) `schemaContent()` helper should be completely removed from RAML 1.0 AST. It was created to help support 0.8 parser users so they could be getting schema contents easily. This helper was created to work with schemas, not something, which is using "schema" keyword, but is actually a type, or is a schema defined via `types` top-level node. We state that schemas and types are totally the same? Very well, then those should be handled uniformly: via the code, which analyses types. Looking at what Juan suggested, that's what he expects.

3) In the AST, we get `schema` and `type` keyword **behaviour** being equal. 
Meaning that if we have the following code:
```RAML
        body:
          application/json:
            type: Person
```
producing:
```json
              "body": {
                "application/json": {
                  "displayName": "application/json",
                  "name": "application/json",
                  "type": [
                    "Person"
                  ],
                  "repeat": false,
                  "required": true
                }
              }
```
Then the following code:
```RAML
        body:
          application/json:
            schema: Person
```
produces:
```json
              "body": {
                "application/json": {
                  "displayName": "application/json",
                  "name": "application/json",
                  "schema": [
                    "Person"
                  ],
                  "repeat": false,
                  "required": true
                }
              }
```

And it does not matter, where `Person` is defined: in `types` or in `schemas` and it does not matter if `Person` is an object type, or if `Person` is an external type (schema).

**TLDR**: compared to what Juan suggested:
Yes, we remove 
```json
"type": [
  "object"
]
```
from Cleaner example. 
Yes, we remove `schemaContent`.
No, we do not rename `schema` to `type` in AST, because those two words have different letters in AST, and we are ruled by an assumption that original code must be always recognizable from AST. So Cleaner example will look like this:
```json
{
  "application/json": {
    "displayName": "application/json",
    "name": "application/json",
    "schema": [
      "Cleaner"
    ],
    "repeat": false,
    "required": true
  }
}
```

**How Juan's algorithm will look now?:**
This is the transformation of the algorithm above to take the changes into account, I am not making any statements regarding the validity of the algorithm, for now.

We assume that `display as type` is known function and it is able to display type whether it is object type, union type, have complex type hierarchy, is external type (schema), or any other case allowed in RAML:
```
if (node has schema property or node has type property)
  var typeValue = schema or type property value, whichever exists
  if (typeValue is null) return;

  if (typeValue is a key in types top section)
    display the entry in types top section as type
  else if (typeValue is a key in schemas top section)
    display the entry in schemas top section as type
  else if (node has properties property)
    display node as type
  else
    display typeValue as string
```

**FAQ, starting with answers regarding algorithm questions from Juan:**

Q: First, will this cover all possible cases?
A: No, not at all.

Q: Second, does anyone else find this very complex?
A: It is not complex enough to cover all possible cases. RAML is much more complex than that. And we should not forget to consider the complexity of `display as type` method itself, there is a lot to handle there.

Q: What is one of the conceptual shortcomings of the algorithm described above?
A: `"application/json"` node here is a type itself, always, whether it has "properties" inside or not. In example, it can have no properties, but have some facets fixed. 
Consider the following RAML code:
```RAML
types:
  SuperTypeWithFacetDefined:
    facets: 
      aRandomFacetPropertyYouCantKnowBeforehandsWithoutAnalyzingAncestor: number
  
...
      206:
        body:
          application/json:
            type: SuperTypeWithFacetDefined
            aRandomFacetPropertyYouCantKnowBeforehandsWithoutAnalyzingAncestor: 5
```
So instead of the algorithm above, the algorithm should always look like:
```
display body as type
```

Q: What is another conceptual shortcoming of the algorithm described above?
A: Type can be defined in a library. Searching in top-level `types` or `schemas` of the current RAML unit is not enough. Also, type defined in a library can extend a type defined in yet another library. Etc.

Q: So, it looks like `display as type` function mentioned above should be itself able to do everything this algorithm handles, and much more, and also do it recursively, handling type hierarchies, aggregation, different kinds of types, and considering different RAML units? 
A: Yes.

Q: Is that possible to do using top-level AST serialized to JSON?
A: Yes, but the pure AST (whether it is serialized to JSON or not) is not the right tool for the job, so this will require to develop a noticeable subset of what we call "type system", and this will be a lot of work.

Q: Is there any way to reduce the cost?
A: Yes, to use the mechanism called "runtime types", which is a subset of type system, and which we use in the parser itself and API Workbench.

Q: Are "runtime types" JSON? 
A: No, because to use that mechanism you need to call methods.

Q: Is that mechanism exposed from the parser for external users?
A: Yes, it is exposed, documented, and there is even Getting Started. This mechanism was many times proposed to use instead of the JSON-serialized AST.

Q: Will using that "runtime types" make developing `the algorithm needed to handle a body node just for displaying it` an easy task taking 10-50 lines of code?
A: Most probably, no. It depends on how detailed we want the types to be displayed. If we just want to handle the schema case, array case, and for object types to only gather and display all type properties (including the ones defined in ancestor, ancestors of ancestor, etc), then probably the algorithm is not going to be that large. Otherwise, if we want to display the hierarchy in details, its still going to be large. 

Q: Is there a sample of how "runtime types" help me displaying the types?
A: Sure. For the RAML samples above, which refer schemas (whether those schemas are referred using `schema` keyword or `type` keyword, whether those schemas are defined at top level using `schemas` keyword or `types` keyword), the following code displays schema contents:
```js
    var responses = api.resources()[0].methods()[0].responses();
    responses.forEach(function(response){
        console.log("Response " + response.code().value() + ":");
        if (response.body()[0] && response.body()[0].runtimeType().hasExternalInHierarchy()) {
            var schemaContents = response.body()[0].runtimeType().externalInHierarchy().schema();
            console.log("Schema:\n" + schemaContents + "\n");
        }
    })
```
The output is:
```
Response 199:
Response 200:
Response 201:
Schema:
{
  "type": "object",
  "properties": [
    {
      "name": "string"
    }
  ]
}

Response 202:
Schema:
{
  "type": "object",
  "properties": [
    {
      "name": "string"
    }
  ]
}

Response 203:
Response 204:
Schema:
{
  "type": "object",
  "properties": [
    {
      "name": "string"
    }
  ]
}

Response 205:
Schema:
{
  "type": "object",
  "properties": [
    {
      "name": "string"
    }
  ]
}
```

Q: Is there any way to jump from JSON node back to AST node in order to call that `runtimeType()` method for a type AST node?
A: No. Not right now. We were considering developing such a mechanism, in example, by assigning each node an ID, putting that ID into `__metadata__` and providing API node with a helper like `getNodeById()`. But as current parser users were split into the ones using AST itself and having AST nodes at hands, and the ones calling `toJSON()` method of API node and then refusing to do anything with AST itself, we abandoned this idea as unneeded by anyone.

Q: How fast is it possible to develop a mechanism to jump from JSON node to AST node?
A: In a matter of days.

**Comment by juancoen** [comment]
> It looks like our mixture of schemas and types comes to the point of referencing types in schema keyword. I personally dislike this and think we should not be mixing those two to that extent, we've got schema keyword as a legacy support, after all. I would left schema only being able to reference schemas defined using schemas keyword, and only allow defining actual schemas under schemas keyword.

I totally agree with this. Can it be changed? @sichvoge @usarid

**Comment by ddenisenko** [comment]
Juan, it will not really help with the whole complexity of a task for display types. I really recommend to consider using runtime types instead of JSON for this type of analysis.

**Comment by juancoen** [comment]
@ddenisenko At this point, I don't think we have a couple of days to wait for the JSON -> runtime types conversion tool, so we'll need to do our best with what is exposed in JSON

**Comment by usarid** [comment]
We will continue to have `schema` be an exact synonym to `type` and `schemas` be an exact synonym to `types`. How the parser output wishes to indicate that one was used vs the other isn't for the spec to decide, of course. Because they are synonyms, you cannot have both `types` and `schemas` in the same document, nor can you have both `type` and `schema` on the same node.

**Comment by ddenisenko** [comment]
The listed changes implemented in 0.2.3.
